src/Network/GRPC/MQTT/Message/Stream.hs

```diff
@@ -124,23 +121,33 @@ makeStreamReader ::
   io (m (Maybe ByteString))
 makeStreamReader channel options = do
   queue <- atomically newTQueue
-  pure (nextStreamChunk queue)
+  isdone <- newIORef False
+  pure (readStreamQueue queue isdone)
   where
-    nextStreamChunk :: TQueue ByteString -> m (Maybe ByteString)
-    nextStreamChunk queue = do
+    nextStreamChunk :: TQueue ByteString -> IORef Bool -> m ()
+    nextStreamChunk queue isdone = do
       runStreamChunkRead channel options >>= \case
-        Nothing -> pure Nothing
-        Just cs -> do
-          liftIO (mapConcurrently_ (atomically . writeTQueue queue) cs)
-          readStreamQueue queue
-
-    readStreamQueue :: TQueue ByteString -> m (Maybe ByteString)
-    readStreamQueue queue = do
-      isdone <- atomically (isEmptyTQueue queue)
-      if isdone
-        then nextStreamChunk queue
-        else atomically (Just <$> readTQueue queue)
-
+        Nothing -> writeIORef isdone True
+        Just cs -> liftIO (atomically (mapM_ (writeTQueue queue) cs))
+
+    readStreamQueue :: 
+      TQueue ByteString ->
+      IORef Bool -> 
+      m (Maybe ByteString)
+    readStreamQueue queue isdone = do
```

Comment:
@riz0id went over these changes with me on a call because I was missing that `makeStreamReader` is returning an action that is executed later so I was confused about the necessity for both TQueue and IORef, which I understand the reasons for better now.